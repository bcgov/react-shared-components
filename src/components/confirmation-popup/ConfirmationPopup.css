.hide-border {
  border-bottom: 0 none;
}

.larger-font {
  font-size: larger;
}

.padding-left {
  padding-left: 0%;
}
