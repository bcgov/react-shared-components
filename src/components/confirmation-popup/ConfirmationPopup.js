/* eslint-disable react/require-default-props */
import React from "react";
import PropTypes from "prop-types";
import Modal from "react-bootstrap/Modal";
import { Button } from "../button/Button";
import { propTypes } from "../../types/propTypes";

import "./ConfirmationPopup.css";

const generateButton = (onClick, label, styling, testId) => (
  <Button onClick={onClick} label={label} styling={styling} testId={testId} />
);

const ConfirmationPopup = ({
  modal: { show, title, body },
  mainButton,
  confirmButton,
  cancelButton,
}) => {
  return (
    <>
      {generateButton(
        mainButton.onClick,
        mainButton.label,
        mainButton.styling,
        "main-cancel-btn"
      )}

      <Modal show={show} onHide={cancelButton.onClick}>
        <Modal.Header className="bcgov-hide-border mt-3" closeButton>
          <Modal.Title className="mx-auto bcgov-larger-font">
            {title}
          </Modal.Title>
        </Modal.Header>
        <div className="mx-auto">
          <Modal.Body className="bcgov-padding-left">{body()}</Modal.Body>
          <div className="mx-auto mb-5">
            {generateButton(
              confirmButton.onClick,
              confirmButton.label,
              confirmButton.styling,
              "modal-confirm-btn"
            )}
            <p />
            {generateButton(
              cancelButton.onClick,
              cancelButton.label,
              cancelButton.styling,
              "modal-cancel-btn"
            )}
          </div>
        </div>
      </Modal>
    </>
  );
};

export default ConfirmationPopup;

ConfirmationPopup.propTypes = {
  modal: PropTypes.shape({
    show: PropTypes.bool.isRequired,
    title: PropTypes.string.isRequired,
    body: PropTypes.func.isRequired,
  }).isRequired,
  mainButton: propTypes.button,
  confirmButton: propTypes.button,
  cancelButton: propTypes.button,
};
