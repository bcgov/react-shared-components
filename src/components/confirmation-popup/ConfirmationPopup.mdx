import { Meta, Story } from "@storybook/addon-docs/blocks";
import { ConfirmationPopup } from "./ConfirmationPopup";

<Meta title="ConfirmationPopup" component={ConfirmationPopup} />

# Confirmation Popup

Add dialogs to your site for lightboxes or user notifications.

## When to Use

The main use case for a confirmation popup is when you need your users to confirm
an action that may lead to unexpected behaviour if it occurs without confirmation. Examples include:

- Exiting the current site/process
- Cancelling a step of the process
- Anything that causes a loss of progress in the process

## When not to Use

- To display forms for the user to complete
- Displaying non-critical information

## API

| Prop          | Type   | Required | Description                                                             | Default |
| ------------- | ------ | -------- | ----------------------------------------------------------------------- | ------- |
| modal         | object | true     | object describing aspects of the modal                                  |         |
| mainButton    | object | true     | object describing the main button that appears outside the popup        |         |
| confirmButton | object | true     | object describing the confirmation button that appears within the popup |         |
| cancelButton  | object | true     | object describing the cancel button that appears within the popup       |         |

### Modal API

The modal object has its own properties that need to be present within it, as shown below:

| Prop  | Type     | Required | Description                                                  | Default |
| ----- | -------- | -------- | ------------------------------------------------------------ | ------- |
| show  | boolean  | true     | value representing whether the popup should be shown or not  |         |
| title | string   | true     | value for the heading of the popup                           |         |
| body  | function | true     | function rendering some jsx to display within the popup body |         |

### Button API

The mainButton, confirmButton, and cancelButton have their own properties as well, as shown below:

| Prop     | Type     | Required | Description                                                                           | Default |
| -------- | -------- | -------- | ------------------------------------------------------------------------------------- | ------- |
| onClick  | function | true     | function executed on button click                                                     |         |
| label    | string   | true     | text to be displayed on button                                                        |         |
| styling  | string   | true     | list of css classnames to apply styling based on (ex: normal-blue, normal-white, btn) |         |
| disabled | boolean  | false    | indicates whether button is disabled or enabled                                       | false   |

<br />

## Default Confirmation Popup

Used as a confirmation for a user cancelling their process.

<Story name="Default" id="confirmation-popup--default"></Story>

```bash
const body = () => (
  <>
    <p>Your files will not be submitted.</p>
    <p>
      You will be returned to:
      <br />
      <b>Original</b> website
    </p>
  </>
);

const modal = {
  title: "Cancel process?",
  body
};

const modal = {
  title: "Cancel process?",
  body,
};

const mainButton = (onButtonClick) => ({
  label: "Click Me",
  styling: "bcgov-normal-white btn",
  onClick: onButtonClick,
});

const confirmButton = (onButtonClick) => ({
  label: "Yes, cancel my process please",
  styling: "bcgov-normal-blue btn bcgov-consistent-width",
  onClick: onButtonClick,
});

const cancelButton = (onButtonClick) => ({
  label: "No, resume my process please",
  styling: "bcgov-normal-white btn bcgov-consistent-width",
  onClick: onButtonClick,
});

<ConfirmationPopup
  key="popup"
  modal={{ ...modal, show }}
  mainButton={mainButton(onButtonClick)}
  confirmButton={confirmButton(onButtonClick)}
  cancelButton={cancelButton(onButtonClick)}
/>
```
