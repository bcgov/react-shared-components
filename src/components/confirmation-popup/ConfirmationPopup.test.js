/* eslint-disable react/jsx-one-expression-per-line */
import React from "react";
import { mount } from "enzyme";
import EnzymeToJson from "enzyme-to-json";

import ConfirmationPopup from "./ConfirmationPopup";

describe("ConfirmationPopup Component", () => {
  const body = () => (
    <>
      <p>Your files will not be submitted.</p>
      <p>
        You will be returned to:
        <br />
        <b>Original</b> website
      </p>
    </>
  );

  const modal = {
    show: true,
    title: "Cancel process?",
    body,
  };

  const onButtonClick = jest.fn();

  const mainButton = {
    label: "main label",
    styling: "normal-blue btn",
    onClick: onButtonClick,
  };

  const confirmButton = {
    label: "Yes, cancel my process please",
    styling: "normal-blue btn consistent-width",
    onClick: onButtonClick,
  };

  const cancelButton = {
    label: "No, resume my process please",
    styling: "normal-white btn consistent-width",
    onClick: onButtonClick,
  };

  const component = (
    <ConfirmationPopup
      modal={modal}
      mainButton={mainButton}
      confirmButton={confirmButton}
      cancelButton={cancelButton}
    />
  );

  test("Matches the snapshot", () => {
    const subject = mount(component);
    expect(EnzymeToJson(subject)).toMatchSnapshot();
  });
});
