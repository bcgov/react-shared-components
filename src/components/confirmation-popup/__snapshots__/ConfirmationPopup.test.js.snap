// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationPopup Component Matches the snapshot 1`] = `
<ConfirmationPopup
  cancelButton={
    Object {
      "label": "No, resume my process please",
      "onClick": [MockFunction],
      "styling": "bcgov-normal-white btn bcgov-consistent-width",
    }
  }
  confirmButton={
    Object {
      "label": "Yes, cancel my process please",
      "onClick": [MockFunction],
      "styling": "bcgov-normal-blue btn bcgov-consistent-width",
    }
  }
  mainButton={
    Object {
      "label": "main label",
      "onClick": [MockFunction],
      "styling": "bcgov-normal-blue btn",
    }
  }
  modal={
    Object {
      "body": [Function],
      "show": true,
      "title": "Cancel process?",
    }
  }
>
  <Button
    disabled={false}
    hasLoader={false}
    label="main label"
    onClick={[MockFunction]}
    styling="bcgov-normal-blue btn"
    testId="main-cancel-btn"
  >
    <button
      className="bcgov-button bcgov-normal-blue btn"
      data-testid="main-cancel-btn"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      main label
    </button>
  </Button>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    dialogAs={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "ModalDialog",
        "render": [Function],
      }
    }
    enforceFocus={true}
    keyboard={true}
    onHide={[MockFunction]}
    restoreFocus={true}
    show={true}
  >
    <Modal
      autoFocus={true}
      backdrop={true}
      backdropTransition={[Function]}
      containerClassName="modal-open"
      enforceFocus={true}
      keyboard={true}
      manager={
        BootstrapModalManager {
          "containers": Array [
            <body
              class="modal-open"
              style="padding-right: 0px;"
            >
              <div
                class="fade modal-backdrop show"
              />
              <div
                aria-modal="true"
                class="fade modal show"
                role="dialog"
                style="display: block;"
                tabindex="-1"
              >
                <div
                  class="modal-dialog"
                >
                  <div
                    class="modal-content"
                  >
                    <div
                      class="bcgov-hide-border mt-3 modal-header"
                    >
                      <div
                        class="mx-auto bcgov-larger-font modal-title h4"
                      >
                        Cancel process?
                      </div>
                      <button
                        class="close"
                        type="button"
                      >
                        <span
                          aria-hidden="true"
                        >
                          ×
                        </span>
                        <span
                          class="sr-only"
                        >
                          Close
                        </span>
                      </button>
                    </div>
                    <div
                      class="mx-auto"
                    >
                      <div
                        class="bcgov-padding-left modal-body"
                      >
                        <p>
                          Your files will not be submitted.
                        </p>
                        <p>
                          You will be returned to:
                          <br />
                          <b>
                            Original
                          </b>
                           website
                        </p>
                      </div>
                      <div
                        class="mx-auto mb-5"
                      >
                        <button
                          class="bcgov-button bcgov-normal-blue btn bcgov-consistent-width"
                          data-testid="modal-confirm-btn"
                          type="button"
                        >
                          Yes, cancel my process please
                        </button>
                        <p />
                        <button
                          class="bcgov-button bcgov-normal-white btn bcgov-consistent-width"
                          data-testid="modal-cancel-btn"
                          type="button"
                        >
                          No, resume my process please
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </body>,
          ],
          "data": Array [
            Object {
              "classes": Array [
                "modal-open",
              ],
              "modals": Array [
                Object {
                  "add": [Function],
                  "backdrop": <div
                    class="fade modal-backdrop show"
                  />,
                  "dialog": <div
                    aria-modal="true"
                    class="fade modal show"
                    role="dialog"
                    style="display: block;"
                    tabindex="-1"
                  >
                    <div
                      class="modal-dialog"
                    >
                      <div
                        class="modal-content"
                      >
                        <div
                          class="bcgov-hide-border mt-3 modal-header"
                        >
                          <div
                            class="mx-auto bcgov-larger-font modal-title h4"
                          >
                            Cancel process?
                          </div>
                          <button
                            class="close"
                            type="button"
                          >
                            <span
                              aria-hidden="true"
                            >
                              ×
                            </span>
                            <span
                              class="sr-only"
                            >
                              Close
                            </span>
                          </button>
                        </div>
                        <div
                          class="mx-auto"
                        >
                          <div
                            class="bcgov-padding-left modal-body"
                          >
                            <p>
                              Your files will not be submitted.
                            </p>
                            <p>
                              You will be returned to:
                              <br />
                              <b>
                                Original
                              </b>
                               website
                            </p>
                          </div>
                          <div
                            class="mx-auto mb-5"
                          >
                            <button
                              class="bcgov-button bcgov-normal-blue btn bcgov-consistent-width"
                              data-testid="modal-confirm-btn"
                              type="button"
                            >
                              Yes, cancel my process please
                            </button>
                            <p />
                            <button
                              class="bcgov-button bcgov-normal-white btn bcgov-consistent-width"
                              data-testid="modal-cancel-btn"
                              type="button"
                            >
                              No, resume my process please
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>,
                  "isTopModal": [Function],
                  "remove": [Function],
                  "setBackdropRef": [Function],
                  "setDialogRef": [Function],
                },
              ],
              "overflowing": true,
              "style": Object {
                "overflow": "",
                "paddingRight": "",
              },
            },
          ],
          "handleContainerOverflow": true,
          "hideSiblingNodes": true,
          "modals": Array [
            Object {
              "add": [Function],
              "backdrop": <div
                class="fade modal-backdrop show"
              />,
              "dialog": <div
                aria-modal="true"
                class="fade modal show"
                role="dialog"
                style="display: block;"
                tabindex="-1"
              >
                <div
                  class="modal-dialog"
                >
                  <div
                    class="modal-content"
                  >
                    <div
                      class="bcgov-hide-border mt-3 modal-header"
                    >
                      <div
                        class="mx-auto bcgov-larger-font modal-title h4"
                      >
                        Cancel process?
                      </div>
                      <button
                        class="close"
                        type="button"
                      >
                        <span
                          aria-hidden="true"
                        >
                          ×
                        </span>
                        <span
                          class="sr-only"
                        >
                          Close
                        </span>
                      </button>
                    </div>
                    <div
                      class="mx-auto"
                    >
                      <div
                        class="bcgov-padding-left modal-body"
                      >
                        <p>
                          Your files will not be submitted.
                        </p>
                        <p>
                          You will be returned to:
                          <br />
                          <b>
                            Original
                          </b>
                           website
                        </p>
                      </div>
                      <div
                        class="mx-auto mb-5"
                      >
                        <button
                          class="bcgov-button bcgov-normal-blue btn bcgov-consistent-width"
                          data-testid="modal-confirm-btn"
                          type="button"
                        >
                          Yes, cancel my process please
                        </button>
                        <p />
                        <button
                          class="bcgov-button bcgov-normal-white btn bcgov-consistent-width"
                          data-testid="modal-cancel-btn"
                          type="button"
                        >
                          No, resume my process please
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>,
              "isTopModal": [Function],
              "remove": [Function],
              "setBackdropRef": [Function],
              "setDialogRef": [Function],
            },
          ],
          "scrollbarSize": 0,
        }
      }
      onEnter={[Function]}
      onEntering={[Function]}
      onEscapeKeyDown={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onHide={[MockFunction]}
      renderBackdrop={[Function]}
      renderDialog={[Function]}
      restoreFocus={true}
      show={true}
      transition={[Function]}
    >
      <Portal
        containerInfo={
          <body
            class="modal-open"
            style="padding-right: 0px;"
          >
            <div
              class="fade modal-backdrop show"
            />
            <div
              aria-modal="true"
              class="fade modal show"
              role="dialog"
              style="display: block;"
              tabindex="-1"
            >
              <div
                class="modal-dialog"
              >
                <div
                  class="modal-content"
                >
                  <div
                    class="bcgov-hide-border mt-3 modal-header"
                  >
                    <div
                      class="mx-auto bcgov-larger-font modal-title h4"
                    >
                      Cancel process?
                    </div>
                    <button
                      class="close"
                      type="button"
                    >
                      <span
                        aria-hidden="true"
                      >
                        ×
                      </span>
                      <span
                        class="sr-only"
                      >
                        Close
                      </span>
                    </button>
                  </div>
                  <div
                    class="mx-auto"
                  >
                    <div
                      class="bcgov-padding-left modal-body"
                    >
                      <p>
                        Your files will not be submitted.
                      </p>
                      <p>
                        You will be returned to:
                        <br />
                        <b>
                          Original
                        </b>
                         website
                      </p>
                    </div>
                    <div
                      class="mx-auto mb-5"
                    >
                      <button
                        class="bcgov-button bcgov-normal-blue btn bcgov-consistent-width"
                        data-testid="modal-confirm-btn"
                        type="button"
                      >
                        Yes, cancel my process please
                      </button>
                      <p />
                      <button
                        class="bcgov-button bcgov-normal-white btn bcgov-consistent-width"
                        data-testid="modal-cancel-btn"
                        type="button"
                      >
                        No, resume my process please
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </body>
        }
      >
        <BackdropTransition
          appear={true}
          in={true}
        >
          <Fade
            appear={true}
            in={true}
            mountOnEnter={false}
            timeout={null}
            unmountOnExit={false}
          >
            <Transition
              addEndListener={[Function]}
              appear={true}
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              timeout={null}
              unmountOnExit={false}
            >
              <div
                className="fade modal-backdrop show"
                onClick={[Function]}
              />
            </Transition>
          </Fade>
        </BackdropTransition>
        <DialogTransition
          appear={true}
          in={true}
          onEnter={[Function]}
          onEntering={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          unmountOnExit={true}
        >
          <Fade
            appear={true}
            in={true}
            mountOnEnter={false}
            onEnter={[Function]}
            onEntering={[Function]}
            onExit={[Function]}
            onExited={[Function]}
            timeout={null}
            unmountOnExit={true}
          >
            <Transition
              addEndListener={[Function]}
              appear={true}
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              timeout={null}
              unmountOnExit={true}
            >
              <div
                aria-modal={true}
                className="fade modal show"
                onClick={[Function]}
                onMouseUp={[Function]}
                role="dialog"
                style={
                  Object {
                    "paddingLeft": undefined,
                    "paddingRight": undefined,
                  }
                }
                tabIndex={-1}
              >
                <ModalDialog
                  onMouseDown={[Function]}
                >
                  <div
                    className="modal-dialog"
                    onMouseDown={[Function]}
                  >
                    <div
                      className="modal-content"
                    >
                      <ModalHeader
                        className="bcgov-hide-border mt-3"
                        closeButton={true}
                        closeLabel="Close"
                      >
                        <div
                          className="bcgov-hide-border mt-3 modal-header"
                        >
                          <ModalTitle
                            className="mx-auto bcgov-larger-font"
                          >
                            <ForwardRef
                              className="mx-auto bcgov-larger-font modal-title"
                            >
                              <div
                                className="mx-auto bcgov-larger-font modal-title h4"
                              >
                                Cancel process?
                              </div>
                            </ForwardRef>
                          </ModalTitle>
                          <CloseButton
                            label="Close"
                            onClick={[Function]}
                          >
                            <button
                              className="close"
                              onClick={[Function]}
                              type="button"
                            >
                              <span
                                aria-hidden="true"
                              >
                                ×
                              </span>
                              <span
                                className="sr-only"
                              >
                                Close
                              </span>
                            </button>
                          </CloseButton>
                        </div>
                      </ModalHeader>
                      <div
                        className="mx-auto"
                      >
                        <ModalBody
                          className="bcgov-padding-left"
                        >
                          <div
                            className="bcgov-padding-left modal-body"
                          >
                            <p>
                              Your files will not be submitted.
                            </p>
                            <p>
                              You will be returned to:
                              <br />
                              <b>
                                Original
                              </b>
                               website
                            </p>
                          </div>
                        </ModalBody>
                        <div
                          className="mx-auto mb-5"
                        >
                          <Button
                            disabled={false}
                            hasLoader={false}
                            label="Yes, cancel my process please"
                            onClick={[MockFunction]}
                            styling="bcgov-normal-blue btn bcgov-consistent-width"
                            testId="modal-confirm-btn"
                          >
                            <button
                              className="bcgov-button bcgov-normal-blue btn bcgov-consistent-width"
                              data-testid="modal-confirm-btn"
                              disabled={false}
                              onClick={[MockFunction]}
                              type="button"
                            >
                              Yes, cancel my process please
                            </button>
                          </Button>
                          <p />
                          <Button
                            disabled={false}
                            hasLoader={false}
                            label="No, resume my process please"
                            onClick={[MockFunction]}
                            styling="bcgov-normal-white btn bcgov-consistent-width"
                            testId="modal-cancel-btn"
                          >
                            <button
                              className="bcgov-button bcgov-normal-white btn bcgov-consistent-width"
                              data-testid="modal-cancel-btn"
                              disabled={false}
                              onClick={[MockFunction]}
                              type="button"
                            >
                              No, resume my process please
                            </button>
                          </Button>
                        </div>
                      </div>
                    </div>
                  </div>
                </ModalDialog>
              </div>
            </Transition>
          </Fade>
        </DialogTransition>
      </Portal>
    </Modal>
  </Modal>
</ConfirmationPopup>
`;
